hydra-node/test/Hydra/HeadLogicSnapshotSpec.hs

```diff
@@ -0,0 +1,244 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE LambdaCase #-}
+{-# OPTIONS_GHC -Wno-unused-do-bind #-}
+
+module Hydra.HeadLogicSnapshotSpec where
+
+import Hydra.Prelude hiding (label)
+import Test.Hydra.Prelude
+
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+import Hydra.Chain (HeadParameters (..))
+import Hydra.Crypto (sign)
+import Hydra.HeadLogic (
+  CoordinatedHeadState (..),
+  Effect (..),
+  Environment (..),
+  Event (NetworkEvent),
+  HeadState (..),
+  OpenState (OpenState),
+  SeenSnapshot (..),
+  collectEffects,
+  collectState,
+  coordinatedHeadState,
+  defaultTTL,
+  isLeader,
+  update,
+ )
+import Hydra.HeadLogicSpec (inOpenState, inOpenState', runEvents, step)
+import Hydra.Ledger (Ledger (..), txId)
+import Hydra.Ledger.Simple (SimpleTx (..), aValidTx, simpleLedger, utxoRef)
+import Hydra.Network.Message (Message (..))
+import Hydra.Options (defaultContestationPeriod)
+import Hydra.Party (deriveParty)
+import Hydra.Snapshot (ConfirmedSnapshot (..), Snapshot (..), getSnapshot)
+import Test.Hydra.Fixture (alice, aliceSk, bob, bobSk, carol, carolSk)
+import Test.QuickCheck (Property, counterexample, forAll, oneof, (==>))
+import Test.QuickCheck.Monadic (monadicST, pick)
+
+spec :: Spec
+spec = do
+  parallel $ do
+    let threeParties = [alice, bob, carol]
+        Ledger{initUTxO} = simpleLedger
+        envFor signingKey =
+          let party = deriveParty signingKey
+           in Environment
+                { party
+                , signingKey
+                , otherParties = List.delete party threeParties
+                , contestationPeriod = defaultContestationPeriod
+                }
+
+    let coordinatedHeadState =
+          CoordinatedHeadState
+            { seenUTxO = initUTxO
+            , allTxs = mempty
+            , seenTxs = mempty
+            , confirmedSnapshot = InitialSnapshot initUTxO
+            , seenSnapshot = NoSeenSnapshot
+            }
+    let sendReqSn =
+          isJust
+            . find
+              ( \case
+                  NetworkEffect ReqSn{} -> True
+                  _ -> False
+              )
+    let snapshot1 = Snapshot 1 mempty []
+
+    let ackFrom sk vk = NetworkEvent defaultTTL vk $ AckSn (sign sk snapshot1) 1
+
+    describe "Generic Snapshot property" $ do
+      prop "there's always a leader for every snapshot number" prop_thereIsAlwaysALeader
+
+    describe "On ReqTx" $ do
+      prop "always emit ReqSn given head has 1 member" prop_singleMemberHeadAlwaysSnapshotOnReqTx
```

Comment:
This property also holds for AckSn. We should try to adapt it

Reply:
We agreed this can be also done separatelly.